web3/middleware/signing.py

```diff
@@ -0,0 +1,44 @@
+from web3.utils.transactions import (
+    fill_transaction_defaults
+)
+import eth_account
+from hexbytes import HexBytes
+
+from cytoolz import (
+    merge,
+)
+
+
+def key_to_account(web3, key):
+    if isinstance(key, eth_account.local.LocalAccount):
+        return key
+    if isinstance(key, (HexBytes, str, bytes)):
+        return web3.eth.account.privateKeyToAccount(key)
+    try:
+        # TODO Find a better way to check for and handle eth_keys objects
+        return web3.eth.account.privateKeyToAccount(key.to_hex())
+    except AttributeError:
+        raise ValueError("key must be one of the types (...). Was of type {0}".format(type(key)))
+
+
+def construct_in_flight_transaction_signing_middleware(private_key):
+
+    def in_flight_transaction_signing_middleware(make_request, web3):
+        account = key_to_account(web3, private_key)
+
+        def middleware(method, params):
+            if method == "eth_sendTransaction":
+
+                transaction_filler = {
+                    'from': account.address,
+                    'nonce': web3.eth.getTransactionCount(account.address)
+                }
+
+                transaction = merge(transaction_filler, params[0])
+                filled_transaction = fill_transaction_defaults(web3, transaction)
+                signed_transaction = account.signTransaction(filled_transaction)
```

Comment:
If `from` is set in the sending transaction to anything other than the public address of `account`, this is going to crash hard. The safest thing to do is to only sign if the transaction has a `from` value set to the `account` address. Then we just skip empty and mismatched `from` values.

Reply:
Also, make sure you use `eth_utils.is_same_address` if there is any way that the two addresses are in different formats (checksummed/not-checksummed)

Reply:
Right, because we reject non-checksummed addresses, but that validation may be somewhere deeper in the middleware "onion" :) -- depending on where this gets injected.